src/container/internal.tsx

```diff
@@ -45,25 +44,10 @@ export default function InternalContainer({
   const baseProps = getBaseProps(restProps);
   const rootRef = useRef<HTMLDivElement>(null);
   const headerRef = useRef<HTMLDivElement>(null);
+  const { isSticky, isStuck, stickyStyles } = useStickyHeader(rootRef, headerRef, __stickyHeader, __stickyOffset);
   const isRefresh = useVisualRefresh(rootRef);
   const hasDynamicHeight = isRefresh && variant === 'full-page';
-  const { isSticky, isStuck, stickyStyles } = useStickyHeader(rootRef, headerRef, __stickyHeader, __stickyOffset);
-
-  /**
-   * Observe the dynamic height of the full page container variant in
-   * visual refresh and update the AppLayout dynamicOverlapHeight property.
-   */
-  const { setDynamicOverlapHeight } = useContext(AppLayoutContext);
-  const [overlapContainerQuery, overlapElement] = useContainerQuery(rect => rect.height);
-
-  useLayoutEffect(
-    function handleDynamicOverlapHeight() {
-      if (hasDynamicHeight) {
-        setDynamicOverlapHeight(overlapContainerQuery ?? 0);
-      }
-    },
-    [hasDynamicHeight, overlapContainerQuery, setDynamicOverlapHeight]
-  );
+  const overlapElement = useDynamicOverlap({ disabled: !hasDynamicHeight });
```

Comment:
Referencing the above comment on the disabled prop...in this instance we can lift the `!hasDynamicHeight` logic and either have a generic ref or use no ref at all. In which case we can remove the `disabled` property entirely from the hook.

Reply:
The optimization is nice, especially since `<Container />` is widely used. One comment though: If we move the decision about "disabled" to attaching or not attaching a ref, we start to spread logic across multiple places in the render function. I have seen it as a convention to have a `disabled` flag on hooks, since we can not just not call them. I will ask for more feedback.

Reply:
Tried to prototype this suggestion, it did not work in practice

```js
const elementRef = useRef();
const conditionalRef = hasOverlap ? elementRef : null;
useDynamicOverlap(conditionalRef);

<div ref={conditionalRef}>
```

It works when transitioning "disabled -> enabled", but in "enabled -> disabled" the container query does not get removed.

If we want to fix it, we need to change `useContainerQuery` signature, allow conditionally disabling this hook

Reply:
This is an interesting problem. It seems like there are three scenarios:

1. Always add a container query to the internal container DOM and call the setter with a zero value if disabled.
2. Remove the `disabled` prop and have a conditional ref at implementation in this component. This will not _add_ unused container queries to the DOM but it will _leave_ unused container queries on the DOM once they are instantiated.
3. Make the adjustment to the `useContainerQuery` hook so we can remove conditional refs at implementation and unused container queries from the overlap hook.

I don't know enough about the `useContainerQuery` to make a judgement here, though it seems like a better long term option than adding a container query to every instance of the internal container.

What do you both think?

Reply:
I think, the current change does not introduce regressions. Because it is also a fix for a bug report, I suggest we proceed merging the current state.

Possible improvements to make the hook smart and conditional can be done as a follow-up

Reply:
Sounds good to me. 👍